gui/action.cpp

```diff
@@ -1175,6 +1175,69 @@ int GUIAction::ozip_decrypt(string zip_path)
 	return ret;
 }
 
+int GUIAction::keypressed()
+{
+	return TWFunc::Exec_Cmd("/sbin/keycheck", false, true);
+}
+
+int GUIAction::keycheck(std::string zip)
+{
+	string insf, pb_installed;
+	insf = "/tmp/pb/installed";
+	int pbRet;
+
+	if (!TWFunc::Path_Exists(insf))
+		return -2;
+
+	pb_installed = TWFunc::File_Property_Get(insf, "installed");
+#ifdef AB_OTA_UPDATER
+	int ramdisk_patched[2], slot = PartitionManager.Get_Active_Slot_Display() == "A" ? 0 : 1;
+	ramdisk_patched[0] = atoi(TWFunc::File_Property_Get(insf, "boot_a").c_str());
+	ramdisk_patched[1] = atoi(TWFunc::File_Property_Get(insf, "boot_b").c_str());
+#endif
+	pbRet = atoi(pb_installed.c_str());
+	if (pbRet && TWFunc::Path_Exists("/sbin/keycheck"))
+	{
+#ifdef AB_OTA_UPDATER
+		if (ramdisk_patched[slot])
+		{
+			gui_print_color("normal", "* * * * * * * * * * * * * * * * * *\n");
+			gui_highlight("pb_flashed_=* Magisk Patched Ramdisk Detected!!!");
+			gui_msg(Msg(msg::kHighlight, "pb_flashed=* New PBRP Flashed, press Volume {1}")("Up For Flashing Magisk."));
+			gui_msg(Msg(msg::kHighlight, "pb_rb_msg=*Volume {1} to Finish")("Down"));
+			gui_print_color("normal", "* * * * * * * * * * * * * * * * * *\n");
+			pbRet = keypressed();
+			if (pbRet == 42) {
+				TWFunc::SetPerformanceMode(true);
+				if (TWFunc::Path_Exists(pb_installed = DataManager::GetCurrentStoragePath() + "/PBRP/tools/magisk.zip"))
+					flash_zip(pb_installed, &w);
+				else
+					flash_zip("/sdcard/PBRP/tools/magisk.zip", &w);
+				TWFunc::SetPerformanceMode(false);
+			}
+		}
+#endif
+		usleep(50000);
+		gui_print_color("normal", "* * * * * * * * * * * * * * * * * *\n");
+		gui_msg(Msg(msg::kHighlight, "pb_flashed=* New PBRP Flashed, press Volume {1}")("Down to Reboot to Recovery."));
+		gui_msg(Msg(msg::kHighlight, "pb_rb_msg=*Volume {1} to Finish")("Up"));
+		gui_print_color("normal", "* * * * * * * * * * * * * * * * * *\n");
+		pbRet = keypressed();
+		if (pbRet == 41) {
+			gui_highlight("pb_saving_log=Saving Taking log...\n");
```

Comment:
_"Saving Taking log..."_ text is unusual English.
Kindly use _"Preserving Logs..."_ instead.